util/timer/src/timer.rs

```diff
@@ -0,0 +1,253 @@
+}
+
+impl TimerLoop {
+    pub fn new(worker_size: usize) -> TimerLoop {
+        let (tx, rx) = channel();
+        let timers = Arc::new(RwLock::new(HashMap::new()));
+
+        let workers = Workers::new(worker_size, Arc::clone(&timers));
+        let worker_queue = workers.queue.clone();
+
+        let waiter_loop = {
+            thread::Builder::new()
+                .name("Timer waiter".to_string())
+                .spawn(move || waiter_loop(rx, worker_queue))
+                .unwrap()
+        };
+
+        TimerLoop {
+            tx,
+            waiter_loop,
+            timers,
+            workers,
+        }
+    }
+
+    pub fn stop(self) {
+        self.workers.stop();
+        self.tx.send(TimerMessage::Stop).unwrap();
+        self.waiter_loop.join().unwrap();
+    }
+
+    pub fn new_timer<T>(&self, id: TimerId, handler: Arc<T>) -> TimerApi
+    where
+        T: 'static + TimeoutHandler + Send + Sync, {
+        let mut timers = self.timers.write();
+        match timers.entry(id) {
+            Entry::Occupied(_) => panic!("Timer id was already occupied: {}", id),
+            Entry::Vacant(entry) => {
+                entry.insert(handler);
+                return TimerApi {
+                    timer_id: id,
+                    tx: self.tx.clone(),
+                }
+            }
+        };
+    }
+}
+
+pub struct TimerApi {
+    timer_id: TimerId,
+    tx: Sender<TimerMessage>,
+}
+
+impl TimerApi {
+    pub fn schedule(&self, after: Duration, timer_token: TimerToken) {
+        self.tx
+            .send(TimerMessage::Schedule(TimerSchedule {
+                at: Instant::now() + after,
+                timer_id: self.timer_id,
+                timer_token,
+            }))
+            .unwrap();
+    }
+}
+
+struct TimerSchedule {
+    at: Instant,
+    timer_id: TimerId,
+    timer_token: TimerToken,
+}
+
+impl std::cmp::PartialEq for TimerSchedule {
+    fn eq(&self, other: &TimerSchedule) -> bool {
+        self.at.eq(&other.at)
+    }
+}
+
+impl std::cmp::Eq for TimerSchedule {}
+
+impl std::cmp::Ord for TimerSchedule {
+    fn cmp(&self, other: &TimerSchedule) -> std::cmp::Ordering {
+        self.at.cmp(&other.at).reverse()
+    }
+}
+
+impl std::cmp::PartialOrd for TimerSchedule {
+    fn partial_cmp(&self, other: &TimerSchedule) -> Option<std::cmp::Ordering> {
+        self.at.partial_cmp(&other.at).map(|x| x.reverse())
+    }
+}
+
+enum TimerMessage {
+    Schedule(TimerSchedule),
+    Stop,
+}
+
+fn waiter_loop(rx: Receiver<TimerMessage>, worker_queue: Arc<WorkerQueue>) {
+    let mut heap = BinaryHeap::new();
+    loop {
+        for msg in rx.try_iter() {
+            match msg {
+                TimerMessage::Stop => return,
+                TimerMessage::Schedule(schedule) => heap.push(schedule),
+            }
+        }
+
+        let wait_for = handle_timeout(&worker_queue, &mut heap);
+
+        match wait_for {
+            Some(timeout) => match rx.recv_timeout(timeout) {
```

Comment:
`recv_timeout` has an issue which causes panic.
https://github.com/rust-lang/rust/issues/39364

We can see it frequently when running the Agent without the Agent Hub.

Reply:
When you visit the issue, someone says he/she is using crossbeam's channel instead of std library.

```
Sending two message was still causing issues and I've refactored my code base to use crossbeam-channels instead: https://docs.rs/crossbeam-channel/0.2.6/crossbeam_channel/#how-to-try-sendingreceiving-a-message-with-a-timeout

So far I haven't had any issues, I would recommend this as a solution to everybody else who runs into this bug.
```

In the agent, I'll try crossbeam's channel.